### PR TITLE
feat(tekton): add runtime value parameter merging functionality

### PR DESCRIPTION
**Pull request by l-qing**
- Add MergeDefaultWithRuntimeValue for updating parameter defaults with runtime values
- Add MergeMapPreserveEmpty to preserve map states during merging
- Support object parameter property-constrained merging
- Include comprehensive test coverage with edge cases
- Handle type mismatches and nil value scenarios gracefully

This enhancement enables dynamic parameter default value updates while maintaining type safety and constraint validation for template parameter processing.

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [`spec` PR link](https://github.com/katanomi/spec) included
- [x] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [x] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [x] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [ ] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

### CODE REVIEW

File: tekton/testdata/param_metadata/spec_merger_edge_cases/empty_destination/destination.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Test case: Empty destination array handling
+# Test Intent: Validates behavior when destination parameter array is empty
+# Key validation points:
+# 1. Empty destination array should be handled gracefully without errors
+# 2. No processing should occur since there are no parameters to merge
+# 3. Result should remain empty even when source parameters exist
+
+# Empty destination parameter specs array
+[]
```
Comment (alaudabot) [inline tekton/testdata/param_metadata/spec_merger_edge_cases/empty_destination/destination.yaml:9]: **[EOF Newline]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing newline

File: tekton/testdata/param_metadata/spec_merger_edge_cases/nil_scenarios/destination.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# 3. Nil to non-nil default transitions should work correctly
+
+- name: param-with-nil-default
+  type: string
+  description: "Parameter with nil default value"
+  # default field intentionally omitted to test nil handling
+
+- name: param-no-default
+  type: array
+  description: "Parameter without any default value"
```
Comment (alaudabot) [inline tekton/testdata/param_metadata/spec_merger_edge_cases/nil_scenarios/destination.yaml:15]: **[EOF Newline]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing newline

File: maps/merge.go [hunk 3]
```diff
@@ -71,3 +71,16 @@ func MergeMapMap(left, right map[string]map[string]string) map[string]map[string
...
+
+// MergeMapPreserveEmpty merges the right map into left map, preserving the original left when both maps are empty.
+// Unlike MergeMap, this function returns the original left map unchanged when both maps are empty,
+// preventing modification of the left map's state (whether nil or empty). If a key exists in both maps,
+// the value from the right map takes precedence.
+// WARNING: This function modifies left in-place when merging is needed.
+// Returns the left map (modified or unchanged) containing all key-value pairs from both maps.
+func MergeMapPreserveEmpty(left, right map[string]string) map[string]string {
+	if len(left) == 0 && len(right) == 0 {
+		return left
```
Comment (l-qing) [inline maps/merge.go:83]: 有些单测中不希望从 map[string]string{} 改为 nil，或者反过来，所以这里在明确不会变更时，保持原样。

File: tekton/cleanup_object_param.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+// CleanupObjectParamByProperties filters object parameter default values based on
+// the parameter's Properties schema definition.
+//
+// This function handles two main scenarios:
+//  1. When Properties is empty: Clears the Default.ObjectVal if it's non-empty,
+//     preserving nil/empty values as-is
+//  2. When Properties is defined: Only retains properties that exist in the Properties
```
Comment (l-qing) [inline tekton/cleanup_object_param.go:29]: 如注释，如果 value 中存在 Properties 中未定义的字段，需要移除掉。

File: tekton/param_metadata.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+// NamedItem defines interface for items that can be identified by name
+type NamedItem interface {
+	GetName() string
+}
+
+// ProcessSelectiveItems processes items selectively, skipping those whose names are in the exclude list.
+// This generic function handles the common logic of filtering items and delegates the actual
+// processing to the ItemProcessor callback.
+func ProcessSelectiveItems[T NamedItem](
```
Comment (l-qing) [inline tekton/param_metadata.go:38]: 本质上就是有选择的处理某些 NamedItem，具体如何处理，由 ItemProcessor 实现。

比如 Build 就是只用改 Tekton.ParamSpec 里面的字段，而 Delivery 还需要改 Artifact 字段。

File: tekton/param_metadata.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+			logger.Debugw("item not processed", "item", itemName)
+		}
+	}
+}
+
+// NewSpecMerger creates an ItemProcessor for merging pipev1beta1.ParamSpec arrays
+// This is a convenience function for common use cases where you want to merge specs directly
+// It uses cached maps to avoid repeated array traversals for better performance
+// defaultSource is used for object type parameters to get additional default values
+func NewSpecMerger(
```
Comment (l-qing) [inline tekton/param_metadata.go:79]: 这里为 `pipev1beta1.ParamSpec` 添加了一个默认实现。

File: tekton/param_metadata.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	// Reuse NewSpecMergerForPointers implementation
+	return NewSpecMergerForPointers(pointerDestination, source, defaultSource)
+}
+
+// NewSpecMergerForPointers creates an ItemProcessor for merging pipev1beta1.ParamSpec pointer arrays
+// This function allows direct modification of the original ParamSpec objects through pointers,
+// avoiding the need for data copying and synchronization
+func NewSpecMergerForPointers(
```
Comment (l-qing) [inline tekton/param_metadata.go:103]: 由于 Delivery 的 spec 也需要改，所以还得支持指针类型 `*[]*pipev1beta1.ParamSpec` 便于再原数组上修改值。

File: tekton/param_metadata.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+		if sourceSpec.Default != nil {
+			(*destination)[destinationIndex].Default = sourceSpec.Default.DeepCopy()
+		}
+
+		// Handle object parameter special processing
+		if (*destination)[destinationIndex].Type == pipev1beta1.ParamTypeObject {
+			// For object type, merge default values from defaultSource if available
+			if defaultSpec, defaultExists := defaultSpecMap[itemName]; defaultExists && defaultSpec.Default != nil {
+				// Merge object defaults: source takes precedence, but preserve defaultSource-only keys
+				(*destination)[destinationIndex].Default = MergeObjectDefaults(defaultSpec.Default, (*destination)[destinationIndex].Default)
```
Comment (l-qing) [inline tekton/param_metadata.go]: @这里在合并用户编排的资源 Build、Delivery 上的 Param 定义，与模板中的 Param 定义做融合。

如果是 object 类型，要额外关注：

- 默认值的合并。
```yaml
# Template（定义安全边界）
- name: git-config
  type: object
  properties: {url: {}, branch: {}}
  default: {url: "template-url", branch: "main"}

# BuildRun（运行时部分覆盖）
- name: git-config
  value: {branch: "feature", illegal: "will-be-filtered"}

# Golden（最终安全结果）
default: {url: "template-url", branch: "feature"}
```

File: tekton/param_metadata.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+				logger.Debugw("merging param object defaults (pointer version)", "param", itemName)
+			}
+
+			// Cleanup default values based on properties schema
+			CleanupObjectParamByProperties((*destination)[destinationIndex])
```
Comment (l-qing) [inline tekton/param_metadata.go]: 由于 BuildRun（运行时部分覆盖） 里面可能传递了多的属性，但改属性模板并不会认，所以需要移除掉。

注意：
- 这个只影响模板渲染时产生的最终 BuildSpec
- **这个不影响 用户自定义 Build 对 TemplateRef 中的赋值操作**

File: tekton/runtime_value_updater.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+	pipelinev1beta1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+// MergeDefaultWithRuntimeValue merges the original ParamSpec default value with runtime value.
+// For object type parameters, only properties defined in original.Properties are merged from runtimeValue.
+// For other types, the default value is completely replaced with a deep copy of runtimeValue.
+// The update modifies original directly while keeping runtimeValue unchanged through deep copying.
+// If the types don't match, no modification is made.
+func MergeDefaultWithRuntimeValue(original *pipelinev1beta1.ParamSpec, runtimeValue *pipelinev1beta1.ParamValue) {
```
Comment (l-qing) [inline tekton/runtime_value_updater.go:30]: 这里根据运行时参数（Parameter）的，来覆盖用户定义（ParameterSpec） 中的默认值。

File: tekton/runtime_value_updater.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+	if original.Type != runtimeValue.Type {
+		return
+	}
+
+	switch original.Type {
+	case pipelinev1beta1.ParamTypeObject:
+		original.Default = MergeObjectDefaults(original.Default, runtimeValue)
+		// Ensure only properties defined in original.Properties are retained
+		CleanupObjectParamByProperties(original)
```
Comment (l-qing) [inline tekton/runtime_value_updater.go:43]: 这里又涉及到 object 类型默认值的合并和清理操作。

File: tekton/runtime_value_updater.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+// - If only one parameter is provided, returns a deep copy of it
+// - Returns nil if no parameters are provided
+//
+// Example:
+//
+//	param1 := &ParamValue{Type: "object", ObjectVal: {"key1": "value1", "key2": "old"}}
+//	param2 := &ParamValue{Type: "object", ObjectVal: {"key2": "new", "key3": "value3"}}
+//	result := MergeObjectDefaults(param1, param2)
+//	// result.ObjectVal = {"key1": "value1", "key2": "new", "key3": "value3"}
+func MergeObjectDefaults(defaults ...*pipelinev1beta1.ParamValue) *pipelinev1beta1.ParamValue {
```
Comment (l-qing) [inline tekton/runtime_value_updater.go:65]: 效果如注释。

File: tekton/param_metadata.go (hunk 9 above)
Comment (l-qing) [reply tekton/param_metadata.go]: Shot 2025 09 22 at 10 53 18

**Review by lentil1016 — DISMISSED** [review]

**Review by lentil1016 — APPROVED** [review]
